tech_reports/LLMs/llms.md

```diff
@@ -73,13 +73,175 @@ Authors:
+
+```python
+in0_block_w=1,
+out_subblock_h=1, # Must be divisible by per_core_M
+out_subblock_w=1, # Must be divisible by per_core_N
+```
+- `in0_block_w` should divide evenly into K. Higher is better. 
+- `out_subblock_h` and `out_subblock_w` should divide evenly into M and N respectively. Higher is better. 
+  - When FP32 accumulate is enabled, `out_subblock_h * out_subblock_w` should be <= 4. Otherwise, `out_subblock_h * out_subblock_w` should be <= 8.
+
+```python
+per_core_M=math.ceil(M / 32 / cores_y),  # M / TILE_HEIGHT / Grid_Size
+per_core_N=math.ceil(N / 32 / cores_x),  # N / TILE_WIDTH / grid width
+```
+- These parameters tell the matmul how many tiles of output each core is responsible for. Therefore, divide M and N by 32 (the tile size) and the core grid size. Round up because you may have padding.
+
+```python
+transpose_mcast=False,
+fused_activation=None,
+fuse_batch=False,
+```
+- If this matmul is part of an MLP with an activation, `fused_activation` will tell the kernel which activation to apply.
+- `fuse_batch` should generally be set to `False`.
+
+Since we use matmul 2D for large matmuls, there may be some issues where we run out of L1 just to store intermediate values in the kernel. When this happens, try reducing `in0_block_w` and `out_subblock_h` and `out_subblock_w`.
+
+##### DRAM-Sharded Matmul
+DRAM-Sharded matmul should be used in decode mode, where activations are small and DRAM-bandwidth to read weights is the limiting factor in op performance. This matmul gets its name because rather than having weights interleaved in DRAM, they are sharded across DRAM banks to optimally collocate weights with compute. See the [DRAM-Sharded Matmul](../Saturating_DRAM_bandwidth/Saturating_DRAM_bandwidth.md) writeup for details on the implementation.
+
+We use DRAM-Sharded matmul for all matmuls in decode mode. The activation and output are width-sharded in L1, and the weights are width-sharded in DRAM. 
+
+To use DRAM-Sharded matmul, create your weight memory config with this helper function we created in [`model_config.py`](../../models/demos/llama3/tt/model_config.py):
+
+```python
+weights_memory_config = create_dram_sharded_mem_config(k=K, n=N)
+```
+
+This function takes care of padding weights to fit evenly into the 12 DRAM banks.
+
+You will also have to create a program config. We have another helper function in `model_config.py` which does this for you:
+
+```python
+matmul_program_config = dram_matmul_config(
+  m=M,
+  k=K,
+  n=N,
+  num_cores=core_grid.num_cores,
+)
+```
+
+The `core_grid` should be the same core grid that the activation is width-sharded on. The output will end up width-sharded on this core grid as well. Call the matmul like this:
+```python
+output = ttnn.linear(
+  activation,
+  weights,
+  compute_kernel_config=compute_kernel_config,
+  core_grid=None,
```

Comment:
was this core grid supposed to be None?

Reply:
yeah, the core grid is specified by the program config. I'll remove this line since it's confusing.